### ISSUE TITLE
Minor tasks from asynchronous review

### ISSUE THREAD

**Issue by kayceesrk** [ocamllabs/ocaml-multicore#742]
This is a list of minor tasks from the asynchronous review phase. Major tasks will have their own issues. The multicore team members should feel free to start picking up these tasks and working on it. Please put your GitHub handle beside the task if you've picked up a task to avoid duplicating the work.

- [x] Please move runtime/gen_sizeclasses.ml to tools/ or remove it entirely.  As far as I could tell, it's not used in normal builds of the system.
- [x] The function build_frame_descriptors should be static or should be renamed caml_build_frame_descriptors.
- [x] For runtime/lf_skiplist.c I trust Luc's extensive review.  Just one cosmetic suggestion:
```
    (see also William Pugh, "Skip lists: a probabilistic alternative to
    balanced binary trees", Comm. ACM 33(6), 1990). */
```
Replace with
```
    A sequential implementation of skip lists is in file skiplist.c and is based on the paper by
    William Pugh, "Skip lists: a probabilistic alternative to
    balanced binary trees", Comm. ACM 33(6), 1990). */
```
- [x] caml_extern_state could be allocated on first use of the marshaling functions in a domain (and then stored in Caml_state).  This would make domain creation faster in the common case where a domain will not use the marshaler.
- [x] runtime/frame_descriptors.[ch] need copyright headers.
- [x] The function build_frame_descriptors should be static or should be renamed caml_build_frame_descriptors.
- [x]  indentation in caml_process_pending_signals_exn needs fixing.
- [x] sync.c  needs copyright header
- [x] sync.c line 191  line too long

**Comment by damiendoligez** [comment]
- [x] Revisit the capping of work-to-do. https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/major_gc.c#L411 https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/major_gc.c#L420 In the sequential GC, the work was capped because you can't do more than 1/3 of a GC cycle in one slice. This turned out to cause problems. Here, you have an accumulated work-to-do, which is not subject to the constraint, so I think it's better not to cap it (never mind doing it twice...). I think it would be better to cap `computed_work` here: https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/major_gc.c#L471
- [x] This debug code looks strange. https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/major_gc.c#L530-L532 If `v` is `Debug_free_major` (or any `Debug_`* value) you probably want a fatal error right there, instead of ignoring the value. Moreover, the `Debug_`* values are odd, so they can't be blocks. ( @sadiqj )
- [x] Misleading comment: https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/major_gc.c#L627-L629 The function (`caml_shrink_mark_stack`) is also called by `domain_terminate` via `caml_finish_marking`. ( @sadiqj )
- [x] I don't understand this comment: https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/major_gc.c#L688 In fact, it seems to me that the header is not accounted for, in this case. ( @sadiqj )
- [x] Let's simplify the code by removing the naked pointer checker (this will need a consensus decision at the synchronous meeting).

**Comment by damiendoligez** [comment]
About `shared_heap.c`

- [x] This is relying on the initial value of mapped memory: https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/shared_heap.c#L172 I'd rather initialize the first pool explicitly like the others.
- [x] `Assert` vs `CAMLassert`: we should choose one and get rid of the other. We'll decide which one at the meeting.
- [x] In this file, you are using `0` as a pointer value instead of NULL. I don't like it and neither does Xavier [citation needed]
- [x] This doesn't work with the way `large_allocate` is called in `caml_shared_try_alloc`: https://github.com/ocaml-multicore/ocaml-multicore/blob/a3832eeaae08ec103d2b4377b5624c081e2782c2/runtime/shared_heap.c#L394 Since this function can be called from the minor GC, this line should be replaced with `if (!a) return NULL;`.
- [x] In `pool_sweep`, when you find a garbage object and add it to the free list, in DEBUG mode you should set its fields to `Debug_free_major`.

**Comment by xavierleroy** [comment]
> In this file, you are using 0 as a pointer value instead of NULL. I don't like it and neither does Xavier [citation needed]

I can live with 0 but I do find NULL clearer.  Also, since NULL is `(void *) 0`, it is typed a little more precisely than 0.  E.g. mistakenly passing NULL to a function expecting a double is an error, but passing 0 is fine.

**Comment by damiendoligez** [comment]
In finalise.c :
- [x] Please add a comment to `caml_final_merge_finalisable` explaining why the young of the source are added to the old of the target. I think I know why but I'd like to be sure.

**Comment by kayceesrk** [comment]
- [x] runtime/amd64.S:589: `jl` (jump if signed less) should be `jb` (jump if unsigned less), since you're comparing pointers.

In asmcomp/amd64/emit.mlp:

- ~[ ] The `preproc_fun` function looks independent of the target processor, so eventually it should be moved to a target-independent module, e.g. Emitaux.  But actually I'd prefer to refactor all the computations of various aspects of functions (whether they contain calls or not, initial stack size, maximal stack size, ...) to after Linearize and before Emit.  (We had this discussion for poll point insertion already.)~ (moved to #754)
- ~[ ] Stack checking is done at the beginning of the function (in Emit.fundecl) while the rest of the function prologue is emitted by the Lprologue instruction since this PR: https://github.com/ocaml/ocaml/pull/2055 , for reasons related to "debugger support".  (Say no more.)  I would prefer all the prologue code to be emitted in one place.  This could mean reverting PR#2055 and removing the Lprologue instruction.~ (moved to #754)
- [x] 838: you moved the cfi_adjust before the I.push instruction (twice), while in the trunk we adjust after push instructions.  Which is best?
- [x] 923:    let (overflow,ret) = new_label(), new_label() in
Inconsistent parentheses + too complicated:
  let overflow = new_label() and ret = new_label() in
- [x] 934: (match !handle_overflow with None -> () | Some (overflow,ret) -> begin ... end);
Would look better as
  begin match !handle_overflow with
  | None -> ()
  | Some (overflow, ret) -> ...
  end;

In runtime/amd64.S:

- [x] Speaking of CFI: could we assume that CFI is always supported?  is it on macOS?  This would avoid a mess of #ifdefs.
- [x] 61: the section name is .text.caml.##name in the current trunk, maybe a recent change that needs to be reflected.
- [ ] 123:  #include "../runtime/caml/config.h" makes me nervous; what is needed exactly? could we include smaller files like m.h (already included at the beginning of amd64.S) and something else to be discussed?
- [ ] 153 and following: could we have C-like pseudocode in comments for what the stack switching operations are doing?  It's kind of hard to follow right now.
- [x] 346 and following: #define for field offsets: move them earlier, before they are used?  mention where the corresponding C structs are defined?
- [x] 397: it looks like the bottom word of a gc_regs_bucket structure sometimes contains the saved value of rax and sometimes contains a pointer to the previous structure.  What about using distinct slots for these two purposes?  It's not like the size of the gc_regs_bucket structure matters.
- [x] 467: caml_call_realloc_stack: not sure if all registers need saving, but it doesn't hurt.
- [x] 596: /* Pop arguments back off the stack */  only needed in DEBUG mode?
- [x] 619-623: movq $0, ...  is this debug code?  should it be in an IF_DEBUG section, then?
- [x] 502, 512, 522: LBL(caml_alloc{1,2,3}) is not useful, as far as I can see.
- [x] 984: wrong comment
        .quad   2           /* three descriptors */

**Comment by xavierleroy** [comment]
I would suggest to delay the following two suggestions for after the MVP (i.e. after merge of Multicore 5.00 in the main OCaml trunk), because they are not urgent at all.

> - The preproc_fun function looks independent of the target processor, so eventually it should be moved to a target-independent module, [...] 
> - Stack checking is done at the beginning of the function (in Emit.fundecl) while the rest of the function prologue is emitted by the Lprologue instruction since this PR

I'd be happy to propose PRs for these two after the MVP merge.

**Comment by kayceesrk** [comment]
Thanks @xavierleroy. I've moved them to a separate issue #754.

**Comment by xavierleroy** [comment]
- [x] In runtime/intern.c: 
https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/intern.c#L489-L490
Unless I'm mistaken, there is no need to initialize the fields to Val_unit, `caml_alloc` already did it.

**Comment by xavierleroy** [comment]
- [x] In asmcomp/selectgen.ml, method `effects_of`:
https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/asmcomp/selectgen.ml#L369
We need to figure out the appropriate effect/coeffect for `Cdls_get`

**Comment by damiendoligez** [comment]
- [x] in runtime/interp.c: https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/interp.c#L557 Resetting the `sp` field of a stack block that you're going to free right away.

**Comment by damiendoligez** [comment]
- [ ] `stack_size_bucket` will only return something if `wosize` is exactly a power of two times `caml_fiber_wsz` but `alloc_stack_noexc` seems to call it with a weaker assumption. https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L87 https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L142
- [x] I don't see why the debug code should be different here, and this seems to break `stack_size_bucket`'s assumptions. https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L388-L390
- [x] Should be `#ifdef DEBUG` https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L135
- [x] The comments or the code seems to need fixing here: https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L280-L282
- [x] Should use the `is_block_and_not_code_frag` function defined on line 305 https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L323
- [x] Did I miss something? Is NULL now a valid value despite NO_NAKED_POINTERS? https://github.com/ocaml-multicore/ocaml-multicore/blob/865aabdacb64db24753b9c9268426a8f34751a26/runtime/fiber.c#L493

**Comment by kayceesrk** [comment]
> Did I miss something? Is NULL now a valid value despite NO_NAKED_POINTERS?

null_stk is `Val_ptr(NULL)`. So it is ok.

**Comment by kayceesrk** [comment]
stdlib/fiber.mli:

- [x] line 33, Continuation_alread_taken should be Continuation_already_taken
- [x] line 52, the type "handler" is defined as "('a,'b) handler" but is then used at line 60 as "('b, 'c) handler". I would suggest using the same type variables names in both definitions; otherwise the reader is forced to perform a mental renaming. Same comment about "effect_handler" at lines 63 and 69. Same comment about "handler" at lines 89, 97, 105.

stdlib/fiber.ml:

- [ ] IMHO, this code cannot be understood without comments. I would suggest writing some "big picture" comments at the beginning of the file, plus comments about the meaning of each type and function definition.

**Comment by Octachron** [comment]
- [x] Looking at the stdlib diff with the anchor point, there is a bit of rebasing noise in `stdlib/filename.ml`:  in the definition of `check_suffix_name` at  line 103, `String.ends_with` has been inlined in multicore.

**Comment by xavierleroy** [comment]
- [x] In runtime/domain.c, function `create_domain`: please initialize `domain->c_stack` to NULL.  So that if C code calls `caml_raise` before any OCaml code is run in the domain, e.g. `caml_stat_alloc` that raises `Out_of_memory`, a clean fatal error is reported.
- [x] More generally, let's go over `create_domain` with a fine comb to make sure all the fields of the domain state are initialized, or to add comments explaining why some fields are left uninitialized and why this is safe.

**Comment by kayceesrk** [comment]
> Speaking of CFI: could we assume that CFI is always supported? is it on macOS? This would avoid a mess of #ifdefs.

I was pointed out there is a `--disable-cfi` option https://github.com/ocaml-multicore/ocaml-multicore/pull/454. IIUC we can't remove the #ifdefs.

**Comment by kayceesrk** [comment]
>  467: caml_call_realloc_stack: not sure if all registers need saving, but it doesn't hurt.

Agreed. The C callee saved registers need not be saved. 

In the previous design, the fiber stacks were managed on the OCaml heap. `caml_try_realloc_stack` may invoke the GC. So we needed to save all the registers as potentially all the values may contain OCaml values, and needed to be considered as roots for the GC. 

This is not something that we need immediately as stack realloc is a rare operation. I created an issue for tracking this feature for the future https://github.com/ocaml-multicore/ocaml-multicore/issues/772.

**Comment by kayceesrk** [comment]
>  596: /* Pop arguments back off the stack */ only needed in DEBUG mode?

`SWITCH_C_TO_OCAML` expects the `rsp` to point to the `c_stack_link` on the C stack. Hence, popping arguments is necessary.

**Comment by kayceesrk** [comment]
> #include "../runtime/caml/config.h" makes me nervous; what is needed exactly? could we include smaller files like m.h (already included at the beginning of amd64.S) and something else to be discussed?

@xavierleroy the only use of this is for the debug-only function `caml_assert_stack_invariants` https://github.com/ocaml-multicore/ocaml-multicore/blob/5.00/runtime/amd64.S#L969. 

I could duplicate the definitions of macros in `amd64.S`, but it may end up difficult to keep them in sync if we need to change them since it would have to be changed for each of the different architectures. 

The other option is to inline the function in `amd64/emit.mlp`, the only place where this is function is called. This function will have to be implemented differently for each architecture anyway. We already have `Config.stack_threshold` being used in `amd64/emit.mlp`. We'd have to add `Config.stack_ctxt_wosize`. 

Any other ideas to do this better?

**Comment by ctk21** [comment]
> * `stack_size_bucket` will only return something if `wosize` is exactly a power of two times `caml_fiber_wsz` but `alloc_stack_noexc` seems to call it with a weaker assumption.

This is a fair review comment; the code is brittle and looks like it might have bad errors. However I will attempt a defence of the existing situation:
 - if we don't get exactly a power of two times, then the cache is never hit; we fall back to allocate/deallocate
 - it turns out we will (nearly always) be a power of 2 times `caml_fiber_wsz`; new stacks start at `caml_fiber_wsz` and double on reallocation

The defence is weak, however I (at least) couldn't see a fast fix as things would need to be tested and written carefully here. I have filed an issue to pick this up #805
